src/compiler/checker.ts

```diff
@@ -34516,13 +34516,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function checkAssertionWorker(node: JSDocTypeAssertion | AssertionExpression, checkMode: CheckMode | undefined) {
         const { type, expression } = getAssertionTypeAndExpression(node);
-        const exprType = checkExpression(expression, checkMode);
         if (isConstTypeReference(type)) {
             if (!isValidConstAssertionArgument(expression)) {
                 error(expression, Diagnostics.A_const_assertions_can_only_be_applied_to_references_to_enum_members_or_string_number_boolean_array_or_object_literals);
             }
-            return getRegularTypeOfLiteralType(exprType);
+            return getRegularTypeOfLiteralType(checkExpression(expression, checkMode));
         }
+        const links = getNodeLinks(node);
+        links.assertionExpressionType = checkExpression(expression, checkMode);

```

Comment:
By moving the `checkExpression` for the `as const` case into the `if`, it's no longer used in the deferred-check case at all; so you shouldn't *actually* need to cache it like this (which is a tad unsafe, so if it's avoidable, I'd like to avoid it), since you'll either do it once early or once deferred.

Reply:
So what're you thinking @weswigham? Place this in an explicit `else`?

Reply:
Or I guess, keep it in the same as before, but then cache it on links unconditionally?

That makes sense, you're calling it twice in some cases anyway now, right?

Reply:
If I understand correctly, that's what I was doing in https://github.com/microsoft/TypeScript/pull/53261/files/45124b1447f2ff88bdf2f9e65f0e1766a80a70a3, and that's what caused `xstate` checking perf to increase by 10-13% (https://github.com/microsoft/TypeScript/pull/53261#issuecomment-1472613782).

Reply:
@gabritto but aren't you calling `checkExpression` twice now?

Reply:
I guess there's no "checking twice" but there's some code deduplication - `checkExpression` is going to happen anyway. I'd rather we got rid of that and go back to having `exprType`

That said, I don't see why avoiding deferral is a problem in that case.

Reply:
>  I'd rather we got rid of that and go back to having exprType

I wouldn't. The `CheckMode`-independent unconditional cache is unsafe and could lead to weird inconsistencies between signature checking states, which can lead to odd editor/command line error inconsistencies. We do it in one or two other places for perf reasons, but it's to be avoided where possible.